Use Boost Unit-Testing-Framework for Unit Tests

clarkpede opened the pull request:
The article ["Best Practices for Scientific Computing"](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745) recommends that programmers "use an off-the-shelf unit testing library". Specifically, they say:

> Creating and managing tests is easier if programmers use an off-the-shelf unit testing library (5b) to initialize inputs, run tests, and report their results in a uniform way.

Since I'm writing more unit tests for the forcing in the hybrid model, I decided to implement some of my tests using the Boost Unit Testing Framework (UTF).  Here's some details:

+ Boost UTF can be used a header-only library, so it doesn't introduce any extra build problems.
+ Boost is already installed as a system library on my ICES computer (albeit a very old version)
+ More modern versions of Boost are available as lmod system modules on ICES computers.

#### Why use Boost UTF?

+ Multiple tests can share the same variables or use the same setup/teardown.
+ The checks themselves have a much simpler syntax than before.
+ Tests can be grouped into test suites.
+ Multiple related tests can be bundled together in a single file.

#### What If I Don't Want to Use Boost UTF?

You don't have to use it. I've implemented a BUILD_TESTS environmental/preprocessor variable that filters out tests built using Boost UTF.  The code compiles and `make check` runs either way.

#### How do I enable the Boost UTF tests?

After running `./bootstrap`, use the configure option `--enable-testing` to enable tests that use the Boost UTF.  If you want to use a specific version of Boost (such as the version loaded using `module load boost`), include something like:
```
--enable-testing --with-Boost-include=$BOOST_INC
```
You don't need to include a `--with-Boost-lib` option because the tests are set up to use the header-only variant.

clarkpede committed:
Added support for running tests using Boost

As of now, only one test has been converted to use Boost.

clarkpede committed:
Forgot to add configure.ac to last commit

clarkpede committed:
Converted some of the existing tests into Boost tests

clarkpede committed:
Added macros to support older versions of Boost

trevilo reviewed (approved):
All looks good to me.  I've used boost unit before and agree that it will make things a bit easier and cleaner.  So, merge whenever you're ready as far as I'm concerned.

clarkpede committed:
Removes some of the CI tests

Travis CI has a hard time finishing the parallel test run in the
available time.  Some of the continuous adjoint tests are removed in
the hope of getting Travis CI to finish under the time limit.
